modules/gateway/gateway.go

```diff
@@ -62,7 +68,30 @@ func (g *Gateway) Close() error {
 	// clear the port mapping (no effect if UPnP not supported)
 	g.clearPort(g.myAddr.Port())
 	// shut down the listener
-	return g.listener.Close()
+	if err := g.listener.Close(); err != nil {
+		errStrs = append(errStrs, fmt.Sprintf("listener error: %v", err))
+	}
+	// Disconnect from peers.
+	for _, p := range g.Peers() {
+		if err := g.Disconnect(p.NetAddress); err != nil {
+			errStrs = append(errStrs, fmt.Sprintf("peer error: %v", err))
+		}
+	}
+	// Sleep to give time for all go routines to exit. This is necessary because
+	// some go routines write to the logger so we must give them time to exit
+	// before closing the logger.
+	// TODO: block until go routines exit instead of sleeping.
+	time.Sleep(100 * time.Millisecond)
+	// Close the logger. The logger should be the last thing to shut down so that
+	// all other objects have access to logging while closing.
+	if err := g.log.Close(); err != nil {
+		errStrs = append(errStrs, fmt.Sprintf("log error: %v", err))
+	}
+
+	if len(errStrs) > 0 {
+		return errors.New(strings.Join(errStrs, "; "))
+	}
```

Comment:
Hmmm I used this pattern a few places in this PR but looking at it again I'm not sure I like it. I also haven't seen any other packages use a similar pattern, although I haven't looked very hard. What do you think about returning just the first error?

```go
var returnErr error // or returnErr could be the func's named return value

if err := foo.Close(); err != nil && returnErr == nil {
	returnErr = err
}
if err := bar(); err != nil && returnErr == nil {
	returnErr = err
}

return returnErr
```